src/main/java/com/mongodb/jdbc/MongoSQLDatabaseMetaData.java

```diff
@@ -266,9 +324,929 @@ public ResultSet getPrimaryKeys(String catalog, String schema, String table)
         throw new SQLFeatureNotSupportedException("TODO");
     }
 
+    private MongoJsonSchema getTypeInfoJsonSchema() {
+
+        MongoJsonSchema schema = MongoJsonSchema.createEmptyObjectSchema();
+        schema.addRequiredScalarKeys(
+                new Pair<>(TYPE_NAME, BsonTypeInfo.BSON_STRING_TYPE_NAME),
+                new Pair<>(DATA_TYPE, BsonTypeInfo.BSON_INT_TYPE_NAME),
+                new Pair<>(PRECISION, BsonTypeInfo.BSON_INT_TYPE_NAME),
+                new Pair<>(LITERAL_PREFIX, BsonTypeInfo.BSON_STRING_TYPE_NAME),
+                new Pair<>(LITERAL_SUFFIX, BsonTypeInfo.BSON_STRING_TYPE_NAME),
+                new Pair<>(CREATE_PARAMS, BsonTypeInfo.BSON_STRING_TYPE_NAME),
+                new Pair<>(NULLABLE, BsonTypeInfo.BSON_INT_TYPE_NAME),
+                new Pair<>(CASE_SENSITIVE, BsonTypeInfo.BSON_BOOL_TYPE_NAME),
+                new Pair<>(SEARCHABLE, BsonTypeInfo.BSON_INT_TYPE_NAME),
+                new Pair<>(UNSIGNED_ATTRIBUTE, BsonTypeInfo.BSON_BOOL_TYPE_NAME),
+                new Pair<>(FIX_PREC_SCALE, BsonTypeInfo.BSON_BOOL_TYPE_NAME),
+                new Pair<>(AUTO_INCREMENT, BsonTypeInfo.BSON_BOOL_TYPE_NAME),
+                new Pair<>(LOCAL_TYPE_NAME, BsonTypeInfo.BSON_STRING_TYPE_NAME),
+                new Pair<>(MINIMUM_SCALE, BsonTypeInfo.BSON_INT_TYPE_NAME),
+                new Pair<>(MAXIMUM_SCALE, BsonTypeInfo.BSON_INT_TYPE_NAME),
+                new Pair<>(SQL_DATA_TYPE, BsonTypeInfo.BSON_INT_TYPE_NAME),
+                new Pair<>(SQL_DATETIME_SUB, BsonTypeInfo.BSON_INT_TYPE_NAME),
+                new Pair<>(NUM_PREC_RADIX, BsonTypeInfo.BSON_INT_TYPE_NAME));
+        return schema;
+    }
+
+    private BsonValue asBsonIntOrNull(Integer value) {
+        if (value == null) {
+            return new BsonNull();
+        }
+        return new BsonInt32(value);
+    }
+
     @Override
     public ResultSet getTypeInfo() throws SQLException {
-        throw new SQLFeatureNotSupportedException("TODO");
+        BsonValue n = new BsonNull();
+        ArrayList<BsonDocument> docs = new ArrayList<>();
+        MongoJsonSchema schema = getTypeInfoJsonSchema();
+
+        docs.add(
+                createBottomBson(
+                        new BsonElement(
+                                TYPE_NAME, new BsonString(BsonTypeInfo.BSON_BINDATA_TYPE_NAME)),
+                        new BsonElement(DATA_TYPE, new BsonInt32(Types.BINARY)),
```

Comment:
should we add a `static getJDBCType` method to `BsonTypeInfo` to get this for each type? I know I also need this info in at least one other place

Reply:
I think we can defer this to the refactor change Natacha is suggesting with making BsonTypeInfo into enums.  The enum would have the TYPE_NAME, DATA_TYPE specified then we can call `enumType.getJdbcType()` such as:
```
       for (BsonTypeInfo currType : BsonTypeInfo.values()) {
           if (currType.isMySqlSupported) {
               docs.add(
                       MySQLDatabaseMetaData.getTypeInfoValuesDoc(
                               currType.getTypeName(),
                               currType.getJdbcType(), //dataType
                               currType.getPrecision(), //precision
                               ... ));
           }
       }
```